### PR TITLE
8288011: StringConcatFactory: Split application of stringifiers

### PR DESCRIPTION
**Pull request by cl4es**
To take optimal advantage of the pre-existing optimization for repeated filters we could split the application of different types of stringifiers.

The resulting difference in order of evaluation is not observable by conventional means since all reference type share the same object stringifier, and the others are filtering primitives (floats and doubles) which have been passed by value already. 

This change neutral on many concatenation expression shapes, but for any complex expressions with interleaving float/double and reference parameters it brings a straightforward reduction in rebinds and underlying LFs generated. For example on the [MixedStringCombinations.java](https://gist.github.com/cl4es/08fb581dece3a73e89bfa52337bc4248) test there's a modest 2% reduction in total classes loaded with this change (from 16209 to 15872)

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.java.net/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8288011](https://bugs.openjdk.org/browse/JDK-8288011): StringConcatFactory: Split application of stringifiers

### Reviewers
 * [Jorn Vernee](https://openjdk.java.net/census#jvernee) (@JornVernee - **Reviewer**) ⚠️ Review applies to [51c841e8](https://git.openjdk.org/jdk/pull/9082/files/51c841e81f36de4dea9786de8c471b3b9211d54b)
 * [Mandy Chung](https://openjdk.java.net/census#mchung) (@mlchung - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/9082/head:pull/9082` \
`$ git checkout pull/9082`

Update a local copy of the PR: \
`$ git checkout pull/9082` \
`$ git pull https://git.openjdk.org/jdk pull/9082/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9082`

View PR using the GUI difftool: \
`$ git pr show -t 9082`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/9082.diff">https://git.openjdk.org/jdk/pull/9082.diff</a>

</details>

### CODE REVIEW

File: src/java.base/share/classes/java/lang/invoke/StringConcatFactory.java [hunk 1]
```diff
@@ -489,17 +491,22 @@ private static MethodHandle generateMHInlineCopy(MethodType mt, String[] constan
...
-                filter = doubleStringifier();
-            }
-            if (filter != null) {
-                if (filters == null) {
-                    filters = new MethodHandle[ptypes.length];
+                if (doubleFilters == null) {
+                    doubleFilters = new MethodHandle[ptypes.length];
                 }
-                filters[i] = filter;
+                doubleFilters[i] = doubleStringifier();
```
Comment (stsypanov) [inline src/java.base/share/classes/java/lang/invoke/StringConcatFactory.java:509]: Why do we have `ptypes[i] = int.class` for `int` and `short`, `ptypes[i] = String.class` for `float`, `double` and `Object` and none for `boolean` and `char`? 🧐

**Review by JornVernee — APPROVED** [review]
This looks good to me, but...

Looking at the code in MethodHandles::filterArguments, and this optimization. It seems that for multiple filters we could potentially always generate just a single new lambda form if we wanted, not just for repeated filters.

It would be similar to `LambdaFormEditor::makeRepeatedFilterForm`. Just need to extend the species data with one new `L_TYPE` field for each unique filter, generate a new getCombiner `Name` for each unique combiner, pass in an array of combiner types instead of just a single one, and pass in an additional array of ints that marks the combiner used by each argument position. Then do similar wiring up as we do now, but calling the corresponding combiner for each argument pos. Right?

(If I'm right about that, maybe that's even a more interesting direction to explore, since it would benefit everyone, and the code in StringConcatFactory could stay the same)

**Comment by cl4es** [comment]
> Looking at the code in MethodHandles::filterArguments, and this optimization. It seems that for multiple filters we could potentially always generate just a single new lambda form if we wanted, not just for repeated filters.
> 
> It would be similar to `LambdaFormEditor::makeRepeatedFilterForm`. Just need to extend the species data with one new `L_TYPE` field for each unique filter, generate a new getCombiner `Name` for each unique combiner, pass in an array of combiner types instead of just a single one, and pass in an additional array of ints that marks the combiner used by each argument position. Then do similar wiring up as we do now, but calling the corresponding combiner for each argument pos. Right?
> 
> (If I'm right about that, maybe that's even a more interesting direction to explore, since it would benefit everyone, and the code in StringConcatFactory could stay the same)

You're right, but I think such a transform would add quite a bit of complexity. 

Back when I added the `makeRepeatedFilterForm` optimization (for the benefit of `StringConcatFactory` 😄 ) I sought to strike a balance between the simple repeatedly-create-`filterArgumentForm`s transform we would do before and more complex forms that would remove more intermediates. I was also a bit concerned that such "shortcutting" transforms could mean we eliminate sharing, and the more complex we make things the harder it gets to reason about. Example:

```java
...
mh1 = filterArgument(baseMH, 0, filter);
mh1 = filterArgument(mh1, 0, filter);

mh2 = filterArgument(baseMH, 0, filter, filter);

mh1.form() == mh2.form(); // false
```

The two different pathways end up with two different LFs, since `makeRepeatedFilterForm` shortcuts. In practice this might be a non-issue, but I figured it'd be nice if we could have a reasonable pathway to resolve mh1.form() and mh2.form() to the same thing in the future.

**Comment by JimLaskey** [comment]
I make heavy use of the `mh = filterArgument(baseMH, 0, arrayOfFilters);` pattern. I assumed it generated fewer LF than making individual calls. Am I wrong?

**Review by mlchung — APPROVED** [review]
Looks good to me.

Like Jorne's observation, it'd benefit everyone if this could generate just a single new lambda form multiple filters, not just for repeated filters if feasible.   Maybe something to explore more in the future with a tradeoff of complexity.

**Comment by cl4es** [comment]
Thanks for reviews. I've filed [JDK-8288738](https://bugs.openjdk.org/browse/JDK-8288738) to explore adding a generalized `LambdaForm` class which would minimize the number of bound arguments and generated classes. I think the changes here are easy enough to back out as part of exploring that route, so I'll go ahead and integrate this enhancement for now.

/integrate

**Pull request closed without merging**